### ISSUE TITLE
Possible false positive warnings

### ISSUE THREAD

**Issue by jafingerhut** [borkdude/clj-kondo#473]
**version**
clj-kondo v2019.09.22-alpha

**platform**
Ubuntu 18.04.3 desktop Linux

**problem**
I am not certain, but it seems like I have found a few warnings that are false positives.  I know this is normal for lint tools, but wanted to at least provide a concrete reproducible example so you can evaluate for yourself whether it is worth the effort of changing clj-kondo.  It might not be.

**repro**
$ git clone https://github.com/clojure/core.rrb-vector
$ cd core.rrb-vector
$ git checkout fbb03b2bcafd0f849754f80de2a6d20bdb00dda0
$ clj-kondo `clojure -Spath` --lint src

**expected behavior**
Most of the many warnings produced I expect.  A few seem like they may be incorrect, listed below.

**actual behavior**
These warnings seem incorrect:
src/main/clojure/clojure/core/rrb_vector/nodes.clj:33:3: error: #'clojure.core/array is private
src/main/clojure/clojure/core/rrb_vector/nodes.clj:44:5: error: #'clojure.core/array is private
src/main/clojure/clojure/core/rrb_vector/nodes.clj:61:5: error: #'clojure.core/array is private
src/main/clojure/clojure/core/rrb_vector/nodes.clj:106:30: info: unresolved symbol objects
src/main/clojure/clojure/core/rrb_vector/rrbt.clj:68:1: warning: IVecImpl already refers to #'clojure.core/IVecImpl
src/main/clojure/clojure/core/rrb_vector/rrbt.clj:76:1: warning: VecSeq already refers to #'clojure.core/VecSeq
src/main/clojure/clojure/core/rrb_vector/rrbt.clj:334:49: info: unresolved symbol objects

The 'array is private' ones are given for a `definterface` form that is defining an interface, not calling clojure.core/array.  The others are defining an implementation for that method, again with no reference to clojure.core/array.

The `unresolved symbol <foo>` warnings are for type tags like `^objects` that appear to me completely correct, and they, or something similarly effective, are necessary to avoid a reflection warning from the Clojure compiler when reflection warnings are enabled.

**Comment by borkdude** [comment]
Thanks for the report. These are all false positives that should be fixed.

**Comment by borkdude** [comment]
@jafingerhut Can you try again with commit https://github.com/borkdude/clj-kondo/commit/1cbaa43080b6de5da83256ef53f9033c50688630?

You can run this project like:

```
clojure -A:clj-kondo --lint /tmp/core.rrb-vector/src 
```

About `VecSeq`, this is part of clojure.core, but only if you require `clojure.gvec`:

https://github.com/clojure/clojure/blob/e78519c174fb506afa70e236af509e73160f022a/src/clj/clojure/gvec.clj#L58

**Comment by jafingerhut** [comment]
I will attempt to attach the complete output I got from the project with the version of clj-kondo and source code core.rrb-vector described in the original report, and the different clj-kondo output I got with the commit you asked about, to this comment.  The first will be called out-kondo1.txt, the second out-kondo2.txt.

If that works, then it seems that the newer output does get rid of some false positive warnings, and at the same time no longer reports warnings about the use of private functions that seem like legitimate warnings.

For example, this line is in output 1, but not output 2:
src/main/cljs/clojure/core/rrb_vector/rrbt.cljs:37:34: error: #'cljs.core/pr-writer is private

Looking at the source code, that warning appears legitimate -- the code is calling cljs.core/pr-writer, and it is declared private in the cljs.core namespace.

[out-kondo1.txt](https://github.com/borkdude/clj-kondo/files/3644428/out-kondo1.txt)
[out-kondo2.txt](https://github.com/borkdude/clj-kondo/files/3644429/out-kondo2.txt)

**Comment by borkdude** [comment]
Thanks. Yeah, that's because currently some macros (like deftype, definterface, cond->) are linted only superficially (e.g. arity linting was turned off and now I also turned off private var linting). This is a trade off between energy invested in implementing a linter for these macros and the benefit returned: how prevalent are these macros in daily programming? As a linter I think it's good to err on the side of false negatives. cond-> is definitely at the top of my list to implement next, deftype, definterface etc might be handled more precisely in the future. Preventing false positives is something that is prioritized over anything else.

**Comment by jafingerhut** [comment]
Sounds good.  I know how tricky it can be to balance false positives, false negatives, and implementation effort.
